### ISSUE TITLE
Level Compensation in Splay and SplayAz seem wrong

### ISSUE THREAD

**Issue by muellmusik** [supercollider/supercollider#5706]
## Environment

* SuperCollider version: 3.11.2
* Operating system: MacOS
* Other details (Qt version, audio driver, etc.):

## Steps to reproduce
**levelComp** is undocumented for some reason in Splay, but in SplayAz says "If true, the signal level is adjusted to maintain overall loudness the same (n.reciprocal.sqrt)". This was added to the help later, and not by the original author, so I'm not sure it's correct. I’m guessing the original intent was to compensate for full scale signals and prevent clipping. Since Splay uses Pan2 internally, and SplayAz uses PanAz I think they should already be equal power so that's not needed unless I've just misunderstood.

This works fine in Splay for control signals (scales by ```n.reciprocal```) but with full range ar signals ```n.reciprocal.sqrt``` scaling will clip.

```supercollider
{ Splay.ar(DC.ar(1) ! 4).poll }.plot
```

SplayAz is both inconsistent and worse, as it does the ```n.reciprocal.sqrt``` scaling to both ar and kr inputs.

## Expected vs. actual behavior

I think all cases of levelComp = true for Splay and SplayAz should scale by n.reciprocal. Unfortunately fixing this would cause noticeable level changes in some existing code, though likely fix some unintentional clipping. It's bad enough that I'd say it's worth the compatibility hit with a prominent warning. A 'classic' mode could be added to easily adjust old code. I guess the other option is to replace the UGens and deprecate, which is not great either.

**Comment by telephon** [comment]
> Since Splay uses Pan2 internally, and SplayAz uses PanAz I think they should already be equal power so that's not needed unless I've just misunderstood.

I always assumed the compensation is for summing the input channels, not for the distribution of the output channels. E.g. if you had a `SplayAz.ar(1, [a, b, c])` you would perhaps not want the signal to be the sum of a, b, c. Or maybe you would?

**Comment by muellmusik** [comment]
> I always assumed the compensation is for summing the input channels, not for the distribution of the output channels. E.g. if you had a `SplayAz.ar(1, [a, b, c])` you would perhaps not want the signal to be the sum of a, b, c. Or maybe you would?

This is what I was trying to say. So we ensure that the signal will not clip if all the signals are full bandwidth. But scaling by n.reciprocal.sqrt does not ensure that. I can imagine cases where you might not want that scaling of course...

**Comment by muellmusik** [comment]
@adcxyz do you know what the original intention was?

**Comment by adcxyz** [comment]
> @adcxyz do you know what the original intention was?

Yes, the idea was to apply the idea of equal power panning by maintaining equal signal power per channel; 
so re-panning say 12 to 5 channels, each of the 5 channels would be at the same volume/power level that each of the indiv. 12 channels had, This seemed more useful to me than linear scaling, though I agree that keeping peak levels the same is also really useful to have. 

I agree that SplayAz kr should really have n.reciprocal scaling. 
Actually, since SplayAz can also be used to scale channels up (distribute few into many channels), 
the equal power by levelComp idea is somehow wrong conceptually already. 
Hmm, will need to think this through better.

**Comment by adcxyz** [comment]
Just saw in SplayAz help file that some of the test cases behave wrongly.
Will look into that ASAP.

**Comment by muellmusik** [comment]
> Actually, since SplayAz can also be used to scale channels up (distribute few into many channels),
> the equal power by levelComp idea is somehow wrong conceptually already.
> Hmm, will need to think this through better.

Okay.


> Yes, the idea was to apply the idea of equal power panning by maintaining equal signal power per channel;
> so re-panning say 12 to 5 channels, each of the 5 channels would be at the same volume/power level that each of the indiv. 12 channels had,

I don't think that's working correctly even in the scale down case? Or?

**Comment by telephon** [comment]
Maybe add a `SplayAz_old` for backward compatibility?

**Comment by adcxyz** [comment]
@muellmusik : 
you are right, the scaling down by sqrt idea is really flawed, and n.reciprocal is a better choice 
- in the worst case, (all signals hard panned to one channel), it is still below clipping. 
```
{ var n =5; Splay.ar(DC.ar(1!5), spread: 0, center: 1, level: 1, levelComp: false).poll }.play;   // [0, 5], linear sum
{ var n =5; Splay.ar(DC.ar(1!5), spread: 0, center: 1, level: 1/n, levelComp: false).poll }.play; // [ 0, 1], 1/n is safe
{ var n =5; Splay.ar(DC.ar(1!5), spread: 0, center: 1, level: (1/n).sqrt, levelComp: false).poll }.play; // [0, 2.36] soso. 
```

But changing the default behavior on Splay.ar will change levels in a lot of code;
on my laptop, I find 370 or so unique scd/schelp files with "Splay.ar". 

So, I wonder whether it is better to keep Splay as it is, 
and add a Splay variant class with the new behavior? 

SplayAz tests and thoughts to follow.

**Comment by telephon** [comment]
Splay is such a nice name, nobody will use the alternative :).

But actually, as I think of it, what is missing in sclang is a systematically clean way to fan in and fan out N –> M channels. We have the bits and pieces, but they are all over the place: `SelectXFocus`, `SplayAz`, `Select`, `PanAz` etc. If we find a way to make this systematic, then I see a chance that a new way could replace the old one.

**Comment by adcxyz** [comment]
- Start of an overview: 
Fan out: 
Pan2, LinPan2: 1 -> 2
PanAz : 1 -> N
Fan in:
Mix : N -> 1
Select, SelectX, SelectXFocus : N -> 1
Splay : N -> 2
Both: 
SplayAz : N -> M

**Comment by telephon** [comment]
a possible interface could be:

`Fan.ar(numChannels, signal, index, spread, interpolation)`

And

`array.fan(numChannels, index, spread, interpolation)`

?

**Comment by muellmusik** [comment]
> So, I wonder whether it is better to keep Splay as it is,
> and add a Splay variant class with the new behavior?

I agree it's difficult to know what to do when it hasn't been working as expected. I suppose another approach would be to change from accepting just true or false on levelComp to allow ```\classic``` mode? And then go through a cycle or warning/deprecation before changing the default?

**Comment by adcxyz** [comment]
Found the bug in SplayAz spreading behavior: 
3 input chans with a spread of 1 currently generates pan values 
[-1, 0, 1], which means chan 0 and 2 coincide; 
[-0.667, 0, 0.667 ] would be correct. 
I think just fixing this with a notice would be fine, 
as the difference most likely has been going unnoticed.

For the levelComp in Splay, allowing more args would work, maybe like this:
```
true // keep with n.reciprocal.sqrt and warn about deprecation 
false // continues as is
// new : 
\linear //  -> n.reciprocal, will be new default after deprecation
\equal, \sqrt // n.reciprocal.sqrt, for code that wants to keep that usage 
```
Also possible, just for discussion (not sure I like it)
a tunable number interface from 0-1, such that
```
0 // no levelComp
0.5 // n.reciprocal ** 0.5, i.e. same as\equal, \sqrt
1 // n.reciprocal ** 1.0, same as \linear
// and anything in between, tunable by ear
```

levelComp in SplayAz maybe needs a bit more redesign than n.reciprocal:
```
10 -> 3 chans n.reciprocal is safe for the worst case of spread 0); for spread 1, level is only 0.42 per chan. 
3 -> 10 chans - n.reciprocal is safe for the worst case (spread 0); for spread 1, loudest level is 0.33 only.
```
Both in Splay and SplayAz help files, we should probably put examples that explain it better, 
and make recommendations, e.g.
1. for maximum safety, use linear, 
2. for roughly equal volume, and old code you don't want to touch, use \equal, 
3. for precise level tuning by hand, turn off levelComp, and put level calc. in the level arg.

**Comment by adcxyz** [comment]
the panpositions calc. fix
```
		var n = max(1, inArray.size);
		//// previous position calc:
		// var pos = if(n == 1) { center } { [ center - spread, center + spread ].resamp1(n) };
		//// fixed so first and last chans do not have the same wrapped pos:
		var spreadMax = spread * (n-1/n); //
		var pos = if(n == 1) { center } { [ center - spreadMax, center + spreadMax ].resamp1(n) };
```

**Comment by telephon** [comment]
I've made a pull request: can you check and maybe post some test code?

We also may want to discuss: this is a breaking change.

**Comment by mtmccrea** [comment]
In summary, there are a few issues here.

`PanAz` was designed based on a closed circle of loudspeakers (**PanAz**imuth), so panning through the endpoints wraps around.  `Splay` seems to incorrectly assume the output channels don't wrap in this way, but rather just terminate (like `Pan2`). @telephon's PR (#6144) fixes this:

> Found the bug in SplayAz spreading behavior:
3 input chans with a spread of 1 currently generates pan values
[-1, 0, 1], which means chan 0 and 2 coincide;
[-0.667, 0, 0.667 ] would be correct.

but it doesn't address the underlying confusion between the two models.

The other two issues are related to normalization, there are some mixed concepts in the discussion. Normalizing signals based on `1/n` and `sqrt(1/n)` is usually for coherent and incoherent signals, respectively (i.e. amplitude-based normalization or energy-based normalization). `Splay` doesn't know anything about the input signals, so it can't make the decision which to use. So `levelCompensation` should be an argument that is a number not a flag, or, better yet, not be done at all and left to the user to precondition the signals sent to it.

But in any case this normalization is incorrectly being applied to the _output_ rather than the input, by being passed to the `level` arg of `PanAz`, which is why the results are counterintuitive, as @rd-- is highlighting above.

Also, has anyone checked whether the `orientation` argument impacts the encoded source locations? I see to recall this impacts the interpretation of `pos` in `PanAz` (another reason `Splay` should probably not be using `PanAz`).

Based on the way `Splay` is documented, it should probably not be using `PanAz`, but rather a  generic VBAP-like algorithm, which would also offer the ability to do this amp-/energy-based scaling with an argument that is a continuous value applied in the panning algorithm itself (using what's called the "p-value", designed for exactly this normalization issue).

`Splay` looks to be a pretty ad-hoc UGen that should probably live in sc3-plugins, but maybe that's another issue...

**Comment by adcxyz** [comment]
@mtmccrea, I could not find a source for p-value, can you recommend one?

Generally, it seems that any change to levelComp is too risky as it will break balance in pieces. 

One way to go would be to recommend in the help file that people turn levelComp off, 
and provide canonical examples for the different common ideas of level compensation. 
Would that make sense?

**Comment by mtmccrea** [comment]
>Generally, it seems that any change to levelComp is too risky as it will break balance in pieces.

Below I'll explain how `levelComp` could be a number (the arg could still allow a boolean for backward compatibility). It won't fully solve this issue, though, so more documentation would help.

>One way to go would be to recommend in the help file that people turn levelComp off,
and provide canonical examples for the different common ideas of level compensation.

The catch with any of this normalization is that the degree to which input signals are interacting (summing) at all in the outputs is dependent on the number of sources and `spread` (and `width` passed to `PanAz`). If you're spreading 5 channels evenly into 15 outputs, the _signals_ don't sum at all. In the stereo case with `Splay`, they're almost always summing to some extent.

So if `levelComp` can be a number in range [0,1] representing a scaling of amplitude normalization (for in-phase signals) energy normalization (for random-phase signals), _and_ still support a boolean for backward compatibility, most cases would be covered.

Then the docs explain this normalization scale and also _warn_ that the caveat is that signals may still clip if multiple sources are distributed with a low `spread` value.

---
@adcxyz 
>I could not find a source for p-value, can you recommend one?

p-value is introduced in this paper:
https://www.aes.org/e-lib/online/browse.cfm?elib=17363
If you're interested but can't access it I can send it along.

The takeaway is that the panning gains can be normalized by amplitude (`1/n`) for in-phase signals, or by energy/power (`1/sqrt(n)`) random-phase signals. 
Formulated in terms of exponents this is 
$$\frac{1}{n^1} \\ \mathrm{and} \\ \frac{1}{n^{1/2}}$$
or, more generically
$$\frac{1}{n^{p}} \\ \mathrm{where} \\ p \in [\frac{1}{2},1]$$

So if `levelComp` were to be a number, it could be normalized for the user between 0 and 1, where 0 corresponds to energy normalization (p=1/2), and 1 is amplitude normalization (p=1).

**Comment by adcxyz** [comment]
... picking this up after a long break ... sigh ... 


Future behaviors of Splay and SplayAz should: 
- be fully backwards compatible
- provide better new options
- for more complex case, explain how to reason about level compensation for the specific case, 
- and provide code examples for the cases discussed.  

new recommended usage: 
dont use built-in level comp, just calc or tune by hand: 
Splay.ar( ... level: myLevel, ... levelComp: false) 
SplayAz.ar( ... level: myLevel, ... levelComp: false) 

// keep old default behavior as is (simple equal power)
Splay.ar( ... level: myLevel, ... levelComp: true) 
SplayAz.ar( ... level: myLevel, ... levelComp: true) 

// new simple way - use numbers for scalable compensation exponent: 
SplayAz.ar( ... level: myLevel, ... levelComp: 0) // flat
SplayAz.ar( ... level: myLevel, ... levelComp: 0.5) // equal power
SplayAz.ar( ... level: myLevel, ... levelComp: 1.0) // equal amplitude sums

In Splay help file: 
give examples how to calculate levels for equal energy vs safe peak ranges, 
including different and possibly dynamic values for spread and center

In SplayAz help file: 
give examples how to calculate levels, based on number of inputs and outputs, 
including different and possibly dynamic values for spread and width 

hope that makes sense?

**Comment by cdbzb** [comment]
This is great.

Small Q: so are you thinking then about levelComp still defaulting to `true` but recommended use being `false`?

If so It might be better to use a new name - perhaps `Spread` that has correct defaults?

**Comment by adcxyz** [comment]
There are not really correct defaults - these are more like preferences for different cases. 
Yes I would leave the Splay levelComp default as is - would break too much code. 

For a new name, I would do that with a fuller reconsideration of how to remix from m to n channels, 
there were several issues starting discussions on that. 

Here is a branch has a first pass for Splay, with levelComp supporting true, false, and 0.0-1.0, 
and lots of examples and tests whiat the output levels are with which settings:
https://github.com/adcxyz/supercollider/tree/Splay_levelComp_flexify

Happy to hear opinions if this is a good way to resolve this for Splay!

As SplayAz is trickier, I will do that separately once Splay has converged to an acceptable resolution.

**Comment by adcxyz** [comment]
@muellmusik, since you raised this, do you think this is an OK resolution?

**Comment by adcxyz** [comment]
Hm, still not really happy to have these details in Splay.help.

It would probably make more sense to write one detailed help/guide page explaining
Level Compensation resp. normalization when changing number of signal channels
This should cover how to consider equal power vs. equal amplitude, 
coherent vs. incoherent signals, possible worst peaks in both cases.
The combinations to play through with examples:
1 -> 2 : Pan2 
N -> 1 : Mix, sum 
N -> 2 : Splay 
1 -> N : PanAz
M -> N : SplayAz

- On the help pages for Splay, SplayAz etc keep explanations short 
and point to Level Compensation page for details. 

Opinions?

**Comment by mtmccrea** [comment]
ok to close with #6804 merged?

**Comment by adcxyz** [comment]
> ok to close with [#6804](https://github.com/supercollider/supercollider/pull/6804) merged?

yes, thanks, closing now.
